src/NuGet.Core/NuGet.ProjectModel/Utf8JsonStreamReader.cs

```diff
@@ -0,0 +1,274 @@
+        private Utf8JsonReader _reader;
+        // The buffer is used to read from the stream in chunks.
+        private byte[] _buffer;
+        private bool _disposed;
+
+        internal Utf8JsonStreamReader(Stream stream) : this(stream, ArrayPool<byte>.Shared.Rent(BufferSizeDefault))
+        {
+
+        }
+
+        internal Utf8JsonStreamReader(Stream stream, byte[] buffer)
+        {
+            if (stream is null)
+            {
+                throw new ArgumentNullException(nameof(stream));
+            }
+
+            _disposed = false;
+            Stream = stream;
+            _buffer = buffer;
+            Stream.Read(_buffer, 0, 3);
+            var offset = 0;
+            if (!_utf8Bom.SequenceEqual(_buffer.AsSpan(0, 3)))
+            {
+                offset = 3;
+            }
+            var blocksRead = Stream.Read(_buffer, offset, _buffer.Length - offset);
+
+            _reader = new Utf8JsonReader(_buffer.AsSpan(0, blocksRead + offset), isFinalBlock: blocksRead + offset < _buffer.Length, state: new JsonReaderState(new JsonReaderOptions
+            {
+                AllowTrailingCommas = true,
+                CommentHandling = JsonCommentHandling.Skip,
+            }));
+            _reader.Read();
+        }
+
+        private Stream Stream { get; set; }
+
+        internal bool IsFinalBlock => _reader.IsFinalBlock;
+
+        internal JsonTokenType TokenType => _reader.TokenType;
+
+        internal int BufferSize()
+        {
+            ThrowExceptionIfDisposed();
+
+            return _buffer.Length;
+        }
+
+        internal bool ValueTextEquals(ReadOnlySpan<byte> utf8Text) => _reader.ValueTextEquals(utf8Text);
+
+        internal bool ValueTextEquals(string text) => _reader.ValueTextEquals(text);
+
+        internal bool TryGetInt32(out int value) => _reader.TryGetInt32(out value);
+
+        internal string GetString() => _reader.GetString();
+
+        internal bool GetBoolean() => _reader.GetBoolean();
+
+        internal int GetInt32() => _reader.GetInt32();
+
+        internal bool Read()
+        {
+            ThrowExceptionIfDisposed();
+
+            bool wasRead;
+            while (!(wasRead = _reader.Read()) && !_reader.IsFinalBlock)
+            {
+                GetMoreBytesFromStream();
+            }
+            return wasRead;
+        }
+
+        internal bool TrySkip()
+        {
+            ThrowExceptionIfDisposed();
+
+            bool wasSkipped;
+            while (!(wasSkipped = _reader.TrySkip()) && !_reader.IsFinalBlock)
+            {
+                GetMoreBytesFromStream();
+            }
+            return wasSkipped;
+        }
+
+        internal string ReadNextTokenAsString()
+        {
+            ThrowExceptionIfDisposed();
+
+            if (Read())
+            {
+                return _reader.ReadTokenAsString();
+            }
+
+            return null;
+        }
+
+        internal string GetCurrentBufferAsString()
```

Comment:
```suggestion
        /// <Summary>For tests only, do not call in production code</Summary>
        internal string GetCurrentBufferAsString()
```

similarly, `BufferLength` is a leaky abstraction for production code.

Even though these methods are internal, it somehow doesn't feel right for these methods to be available to all the other production code in the same assembly.  How horrible would it be to remove both methods, and then use reflection in the test code to get `_buffer`? 😱

Reply:
I couldn't figure out how to get reflection to work on a ref struct. I did switch the constructor to take in an ArrayPool<byte> and updated the tests to pass in mock array pool. Lemme know what you think of that solution.